docs/runbooks/VirtualMachineStuckOnNode.md

```diff
@@ -0,0 +1,265 @@
+
+# Check pod status and events
+kubectl describe pod <virt-launcher-pod> -n <namespace>
+
+# Check pod logs for errors
+kubectl logs <virt-launcher-pod> -n <namespace> -c compute
+kubectl logs <virt-launcher-pod> -n <namespace> -c istio-proxy  # if using Istio
+```
+
+### 3. Investigate Node Health
+```bash
+# Check node status and conditions
+kubectl describe node <node-name>
+
+# Check node resource utilization
+kubectl top node <node-name>
+
+# Check virt-handler on the affected node
+kubectl get pods -n kubevirt -o wide | grep <node-name>
+kubectl logs <virt-handler-pod> -n kubevirt
+```
+
+### 4. Check Storage and Volumes
+```bash
+# Verify PVC status and mounting
+kubectl get pvc -n <namespace>
+kubectl describe pvc <pvc-name> -n <namespace>
+
+# Check volume attachments on the node
+kubectl get volumeattachment | grep <node-name>
+
+# For DataVolumes, check their status
+kubectl get dv -n <namespace>
+kubectl describe dv <dv-name> -n <namespace>
+```
+
+### 5. Verify Image Accessibility from Node
+```bash
+# Check if container images can be pulled on the node
+kubectl debug node/<node-name> -it --image=busybox
+# Then inside the debug pod:
+# crictl pull <vm-disk-image>
+```
+
+### 6. Check QEMU Process (if accessible to node)
+```bash
+# SSH to node and check QEMU processes
+ps aux | grep qemu | grep <vm-name>
+
+# Check libvirt domains
+virsh list --all | grep <vm-name>
+virsh dumpxml <domain-name>
+```
+
+## Mitigation
+
+### Pod-Level Issues
+1. **Restart virt-launcher pod**:
+   ```bash
+   kubectl delete pod <virt-launcher-pod> -n <namespace>
+   # The VMI controller will recreate it
+   ```
+
+2. **Check resource constraints**:
+   ```bash
+   kubectl describe pod <virt-launcher-pod> -n <namespace>
+   # Look for resource limit violations
+   ```
+
+### Image Issues on Node
+1. **Clear image cache** on the problematic node:
+   ```bash
+   # On the node (requires node access):
+   crictl rmi <problematic-image>
+   ```
+
+2. **Force image re-pull**:
+   ```bash
+   # Delete and recreate the virt-launcher pod
+   kubectl delete pod <virt-launcher-pod> -n <namespace>
+   ```
+
+### Storage Issues
+1. **Check PVC binding and mounting**:
+   ```bash
+   kubectl get pvc -n <namespace>
+   # If PVC is stuck, check the storage provisioner
+   ```
+
+2. **Resolve volume attachment issues**:
+   ```bash
+   kubectl get volumeattachment
+   # Delete stuck volume attachments if necessary
+   kubectl delete volumeattachment <attachment-name>
+   ```
+
+### Node-Level Issues  
+1. **Drain and uncordon node** if it's in a bad state:
+   ```bash
+   kubectl drain <node-name> --ignore-daemonsets --delete-emptydir-data
+   kubectl uncordon <node-name>
+   ```
+
+2. **Restart node-level components**:
+   ```bash
+   # Restart virt-handler on the node
+   kubectl delete pod <virt-handler-pod> -n kubevirt
+   
+   # If necessary, restart container runtime (requires node access)
+   systemctl restart containerd  # or crio
```

Comment:
Rather unspecific suggestions if you ask me, why would you do this? This can be harmful to other workloads on the node.

Reply:
Updated